Remove `IRenderedFragmentBase` and `IRenderedComponentBase`

linkdotnet opened the pull request:
## Info
This PR should ideally target `v2` directly. Currently, this is blocked as it builds on top of`v2-merge-projects`. Once this PR is merged, we can target `v2`directly.

## Motivation
This Pull Request removes `IRenderedFragmentBase` and `IRenderedComponentBase` as it is no longer needed (due to the merge). This will also make it easier once we merge the `Render(Component)` methods. 

I explicitly did not merge `IRenderedComponent` and `IRenderedFragment` for a simple reason: Assume we would do this, then the signature would have a `TComponent? Instance` member in the interface. This is nice but shifts the burden to the user rather than the library. I want to keep `RenderFragment`s backed separately by an instance and one without. That makes it easier for users. Of course, this is open for debate, but I am really in favor of not shipping nullable types to the user.

egil commented:
I'll take a deep look at both PRs tonight. Quick thought though: do we need these things to be interfaces? Could just be classes I think .

linkdotnet commented:
> I'll take a deep look at both PRs tonight. Quick thought though: do we need these things to be interfaces? Could just be classes I think .

To make it work a majority of other classes and methods have to be a) exposed as instances not as classes and b) have to define anchor points for inheritance.

Take our ITestRender and the underlying WebTestRenderer.
The user has to register its own derived render into the container just to return their unique RenderedFragment implementation.
By far it is not impossible but it will be much more effort to do things right with inheritance and extensibility points in mind. I am not sure if we have users willing to extend our object rather than just suggesting it directly on our GitHub tracker... but thanks just a feeling and is not really bagged by evidence.

linkdotnet commented:
I merged `WebTestRenderer` and `TestRenderer`.

linkdotnet commented:
I also added a `new()` constraint for the `ComponentFactoryCollectionExtensions.Add` method.
The reason why I added this is because the implementation uses: `Activator.CreateInstance()` that tries to call the parameterless constructor - and fails if there isn't one around.

The change makes this more explicit for the user. I don't expect any problem because if people used a class without a parameterless ctor it would result in an exception. Therefore I did not add it to the `Migration.md` file.

Anyway - Feel free to revert single commits if you wish. All the work are mere ideas I like, nothing set in stone.

egil commented:
Gah, assembling ikea furniture got the best of me. Off to bed. I'll try to review tomorrow.

linkdotnet commented:
> Gah, assembling ikea furniture got the best of me. 

Of whom, it doesn't? ;) Disassembling is most often even worse ;)

egil commented:
> I also added a `new()` constraint for the `ComponentFactoryCollectionExtensions.Add` method.
> The reason why I added this is because the implementation uses: `Activator.CreateInstance()` that tries to call the parameterless constructor - and fails if there isn't one around.

With the new constraint, we can even skip reflection and just use new T() instead.

linkdotnet commented:
> > I also added a `new()` constraint for the `ComponentFactoryCollectionExtensions.Add` method.
> > The reason why I added this is because the implementation uses: `Activator.CreateInstance()` that tries to call the parameterless constructor - and fails if there isn't one around.
> 
> With the new constraint, we can even skip reflection and just use new T() instead.

Oh yeah - I used that as well.

egil commented:
> > I'll take a deep look at both PRs tonight. Quick thought though: do we need these things to be interfaces? Could just be classes I think .
> 
> To make it work a majority of other objects and methods have to be a) exposed as classes and not as interfaces and b) have to define anchor points for inheritance.
> 
> Take our ITestRender and the underlying WebTestRenderer. The user has to register its own derived render into the container just to return their unique RenderedFragment implementation (plus our current extension methods like `Find` make it practically impossible for a user to have a custom `RenderFragment`). By far it is not impossible but it will be much more effort to do things right with inheritance and extensibility points in mind. I am not sure if we have users willing to extend our object rather than just suggesting it directly on our GitHub tracker... but this just a feeling and is not really bagged by evidence.

Good points. I think we could get around a lot of that with small changes to the design, but that can come later.

egil commented:
> > > I also added a `new()` constraint for the `ComponentFactoryCollectionExtensions.Add` method.
> > > The reason why I added this is because the implementation uses: `Activator.CreateInstance()` that tries to call the parameterless constructor - and fails if there isn't one around.
> > 
> > 
> > With the new constraint, we can even skip reflection and just use new T() instead.
> 
> Oh yeah - I used that as well.

Just saw that right after I commented :)

egil reviewed (dismissed):
Looks good to me.

linkdotnet commented:
I rebased the squashed merge - sorry for interrupting. I need a new approval ;)

PS: Thanks to the merge and only one target framework, the build and test time is so swift now!
